File: Data Structures & Algorithms/regular-expression-matching/submission-18.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int n{static_cast<int>(s.size())}, m{static_cast<int>(p.size())};
        vector<vector<int>> dp(n + 1, vector<int> (m + 1, -1));
        auto back = [&](auto self, int i, int j) {
            if (i == n && j == m) return true;
            if (i > n || j > m) return false;
            bool res{false};
            if (dp[i][j] != -1) return dp[i][j] == 1 ? true : false;
            if (j < m && p[j + 1] == '*' ) {
                if (i == n || s[i] == p[j] || p[j] == '.')
                    res = res || self(self, i + 1, j);
                res = res || self(self, i, j + 2);
            }
            else if (p[j] == '.' || s[i] == p[j]) {
                res = res || self(self, i + 1, j + 1);
            }
            dp[i][j] = res ? 1 : 0;
            return res;
        };
        return back(back, 0, 0);
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-18.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxPathSum(TreeNode root) {
        Stack<TreeNode> st = new Stack<>();
        TreeNode node = root;
        TreeNode last = null;
        HashMap<TreeNode, Integer> mp = new HashMap<>();
        mp.put(null, 0);
        int res = Integer.MIN_VALUE;
        while (node != null || st.size() > 0) {
            if (node != null) {
                st.push(node);
                node = node.left;
            }
            else {
                node = st.peek();
                if (node.right == last || node.right == null) {
                    st.pop();
                    int left = mp.get(node.left);
                    int right = mp.get(node.right);
                    left = Math.max(left, 0);
                    right = Math.max(right, 0);
                    res = Math.max(left + right + node.val, res);
                    int help = Math.max(left, right);
                    mp.put(node, help + node.val);
                    last = node;
                    node = null;
                }
                else {
                    node = node.right;
                }
            }
        }
        return res;
    }
}
